src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -0,0 +1,721 @@
+     */
+    private void reportConsumedData() {
+        long totalNumberOfBytesRead = getTotalOffset() + (peekIndex - valuePreHeaderIndex);
+        dataHandler.onData((int) (totalNumberOfBytesRead - lastReportedByteTotal));
+        lastReportedByteTotal = totalNumberOfBytesRead;
+    }
+
+    /**
+     * Advances to the next token, seeking past the previous value if necessary. After return `event` will convey
+     * the result (e.g. START_SCALAR, END_CONTAINER)
+     * @return false if the next token was an Ion version marker or NOP pad; otherwise, true. If false, this method
+     *  should be called again to advance to the following value.
+     */
+    private boolean nextToken() {
+        if (peekIndex < valueMarker.endIndex) {
+            peekIndex = valueMarker.endIndex;
+        } else if (valueTid != null && valueTid.isDelimited) {
+            seekPastDelimitedContainer_1_1();
+        }
+        valueTid = null;
+        if (dataHandler != null) {
+            reportConsumedData();
+        }
+        if (peekIndex >= limit) {
+            setCheckpointBeforeUnannotatedTypeId();
+            if (parent != null && parent.endIndex == peekIndex) {
+                event = Event.END_CONTAINER;
+            }
+            return true;
+        }
+        reset();
+        int b;
+        if (parent == null) { // Depth 0
+            valuePreHeaderIndex = peekIndex;
+            b = buffer[(int)(peekIndex++)] & SINGLE_BYTE_MASK;
+            if (b == IVM_START_BYTE) {
+                readIvm();
+                return false;
+            }
+        } else {
+            if (nextContainedToken()) {
+                return true;
+            }
+            b = buffer[(int)(peekIndex++)] & SINGLE_BYTE_MASK;
+        }
+        if (readHeader(b, false, valueMarker)) {
+            valueTid = valueMarker.typeId;
+            return true;
+        }
+        valueTid = valueMarker.typeId;
+        return false;
+    }
+
+    @Override
+    public Event nextValue() {
+        event = Event.NEEDS_DATA;
+        while (true) {
+            if (nextToken()) {
+                break;
+            }
+        }
+        return event;
+    }
+
+    @Override
+    public Event fillValue() {
+        event = Event.VALUE_READY;
+        return event;
+    }
+
+    @Override
+    public Event getCurrentEvent() {
+        return event;
+    }
+
+    public int getIonMajorVersion() {
+        return majorVersion;
+    }
+
+    public int getIonMinorVersion() {
+        return minorVersion;
+    }
+
+    public boolean hasAnnotations() {
+        return hasAnnotations;
+    }
+
+    Marker getValueMarker() {
+        return valueMarker;
+    }
+
+    /**
+     * Slices the buffer using the given offset and limit. Slices are treated as if they were at the top level.
+     * @param offset the offset at which the slice will begin.
+     * @param limit the slice's limit.
+     */
```

Comment:
What does the slice represent?

Reply:
A "span" of bytes that comprise a value in the stream. See PR #513